bootstrap_test.py

```diff
@@ -280,6 +280,55 @@ def test_consistent_range_movement_true_with_rf1_should_fail(self):
     def test_consistent_range_movement_false_with_rf1_should_succeed(self):
         self._bootstrap_test_with_replica_down(False, rf=1)
 
+    def test_rf_gt_nodes_multidc_should_succeed(self):
+        """
+        Test creating a KS with RF > N on multi DC doesn't break bootstrap
+        @jira_ticket CASSANDRA-16296 CASSANDRA-16411
+        """
+        cluster = self.cluster
+        cluster.set_environment_variable('CASSANDRA_TOKEN_PREGENERATION_DISABLED', 'True')
+        cluster.populate([1, 1])
+        cluster.start()
+
+        node1 = cluster.nodelist()[0]
+        session = self.patient_cql_connection(node1)
+        session.execute("CREATE KEYSPACE k WITH REPLICATION = {'class' : 'NetworkTopologyStrategy', 'dc1' : '3'}")
+        warning = 'Your replication factor 3 for keyspace k is higher than the number of nodes 1 for datacenter dc1'
+
+        if cluster.version() >= '4.0':
+            warnings_total = 0
+            for node in cluster.nodelist():
+                warnings_total += len(node.grep_log(warning))
+            assert warnings_total == 1
+
+        session.execute("ALTER KEYSPACE k WITH REPLICATION = {'class' : 'NetworkTopologyStrategy', 'dc1' : '2'}")
+        session.execute("CREATE TABLE k.testgtrfmultidc (KEY text PRIMARY KEY)")
+        session.execute("INSERT INTO k.testgtrfmultidc (KEY) VALUES ('test_rf_gt_nodes_multidc_should_succeed')")
+        warning = 'Your replication factor 2 for keyspace k is higher than the number of nodes 1 for datacenter dc1'
+
+        if cluster.version() >= '4.0':
+            warnings_total = 0
+            for node in cluster.nodelist():
+                warnings_total += len(node.grep_log(warning))
+            assert warnings_total == 1
+
+        node3 = Node(name='node3',
+                     cluster=cluster,
+                     auto_bootstrap=True,
+                     thrift_interface=('127.0.0.3', 9160),
+                     storage_interface=('127.0.0.3', 7000),
+                     jmx_port='7300',
+                     remote_debug_port='0',
+                     initial_token=None,
+                     binary_interface=('127.0.0.3', 9042))
+        cluster.add(node3, is_seed=False, data_center="dc1")
+        node3.start(wait_for_binary_proto=True);
+        if cluster.version() >= '4.0':
+            assert len(node3.grep_log(warning)) == 0
```

Comment:
this check for warning can be missed because number of nodes changed.
so for example `of nodes 1` will not be found if the warning would be `of nodes 2` (as we added the node).

Reply:
We could just mark the logs before adding the node and search in every node for a generic `'Your replication factor'` message, for example something like:
```python
marks = map(lambda n: n.mark_log(), cluster.nodelist())

# add node...

if cluster.version() >= '4.0':
    warning = 'Your replication factor'
    for (node, mark) in zip(cluster.nodelist(), marks):
        assert len(node.grep_log(warning, from_mark=mark)) == 0
```

Reply:
Amended but I used a more particular substring for the check to avoid false matches.